Added aggregation pipeline operators (#4)
LucasSGomide/MongoDB: README.md

## Aggregation

Aggregation expressions use **field path** to access fields in the input documents. To specify a field path, prefix the field name with a dollar sign $.

# Aggregation Pipeline

> The aggregation pipeline is a framework for MongoDB. Documents enter a multi-stage pipeline that transforms the documents into an aggregated result. Each stage transforms the documents as they pass from one stage to another. 

Reviewer: ```suggestion
The aggregation pipeline is a framework for MongoDB. Documents enter a multi-stage pipeline that transforms the documents into an aggregated result. Documents pass through the stages in sequence. Each stage transforms the documents as they pass from one stage to another. 
```

## Aggregation Operators

### $MATCH

````javascript
db.collection.aggregate([
    { $match: { <query> } },
]);
````

Filters the documents so only those that match the specified condition(s) may enter the next stage. It resembles the filters from **find()** method.

```javascript
db.collection.aggregate([
    { $match: { name: "Alex" } },
]);
```

```javascript
db.collection.find({ name: "Alex" });
```

Select all documents that match the query.

This operator may be used at any stage of the pipeline. When used as the first stage, it should improve considerably the performace of your query. In the other hand, when used at any other point, it will act similar to a **HAVING** from the SQL.

---

### $LIMIT

```` javascript
db.collection.aggregate([
    { $limit: <positive integer> },
]);
````

> Limits the number of documents passed to the next stage in the pipeline.

Reviewer: ```suggestion
Passes only the first n documents to the next stage in the pipeline, in which n is the specified positive integer.
```

````javascript
db.collection.aggregate([
	{ $match: { price: 10 } },
    { $limit: 5 },
]);
````

> Returns five results that matches the specified condition from the previous stage.

Reviewer: ```suggestion
Returns the first five results that matches the specified condition from the previous stage.
```

---

### $LOOKUP

This operator allows the use of documents from another collection (table), similar to a **JOIN** from the SQL. The documents from the joined collection will be avaible through a new array.

There are two types of `$lookup`:

#### Equality Match

````javascript
db.collection.aggregate([
	{
        $lookup:
        {
          from: <collection to join>,
       	  localField: <field from the input documents>,
       	  foreignField: <field from the documents of the "from" collection>,
       	  as: <output array field>
        },
    },
]);
````

````javascript
db.collection.aggregate([
	{
        $lookup:
        {
          from: <inventory>,
       	  localField: <item>,
       	  foreignField: <sku>,
       	  as: <inventory_docs>
        },
    },
]);
````

----

### $GROUP

> This operator groups documents by the specified `_id` expression. It's worth mentioning that the specified `_id` is actually a "parameter" of `$group` and is not to be mistaken with `_id` from the collections.

Reviewer: ```suggestion
This operator groups documents by the specified `_id` expression. It's worth mentioning that the specified `_id` is actually a "parameter" of `$group` and is not to be mistaken with `_id` from the collections. The operator `_id` is required. If you specify a value of `null` the $group stage calculates accumulated values for the input document as a whole.
```

````javascript
db.collection.aggregate([
	{
		$group:
		{
			_id: <expression>,
			<field1>: { <accumulator1> : <expression1> },
			...
			<fieldN>: { <accumulatorN> : <expressionN> },
		},
	},
]);
````

> The `<field>` may be named as you see fit, howerver, the `<accumulator>` must be a accumulator operator.

Reviewer: ```suggestion
The `<field>` may be named as you see fit, however, the `<accumulator>` must be an accumulator operator.
```

````javascript
db.collection.aggregate([
  {
    $group : {
      _id : "$itemId",
      count: { $sum: 1},
    }
  }
]);
````

The output would be something like this:

````bash
{ _id : "Chair", count : 3 }
{ _id : "Table", count : 2 }
{ _id : "Lamp", count : 2 }
````

> **OBS:** If you set `_id` as `null`, you will group all documents.

Reviewer: ```suggestion
```

#### List of most used accumulators:

- $addToSet
- $avg
- $first
- $last
- $max
- $sum

---

### $UNWIND

````javascript
db.collection.aggregate([
    { $unwind: <field path> },
]);
````

This operator desconstructs an array field into documents. To be more precisely, a document for each element of the array.

Let's take the following document as an example:

````javascript
{
    _id: 7,
    name: "GMusicApp",
    subscription: ["free", "premium", "family", "university"]
}
````

> If we were to apply the `$unwind` operator:

Reviewer: ```suggestion
If we apply the `$unwind` operator:
```

## Aggregation

Aggregation expressions use **field path** to access fields in the input documents. To specify a field path, prefix the field name with a dollar sign $.

# Aggregation Pipeline

The aggregation pipeline is a framework for MongoDB. Documents enter a multi-stage pipeline that transforms the documents into an aggregated result. Each stage transforms the documents as they pass from one stage to another. 

## Aggregation Operators

> ### $match

Reviewer 2: ```suggestion
### $MATCH
```

````javascript
db.collection.aggregate([
    { $match: { <query> } },
]);
````

Filters the documents so only those that match the specified condition(s) may enter the next stage. It resembles the filters from **find()** method.

```javascript
db.collection.aggregate([
    { $match: { name: "Alex" } },
]);
```

```javascript
db.collection.find({ name: "Alex" });
```

Select all documents that match the query.

This operator may be used at any stage of the pipeline. When used as the first stage, it should improve considerably the performace of your query. In the other hand, when used at any other point, it will act similar to a **HAVING** from the SQL.

---

> ### $limit

Reviewer 2: ```suggestion
### $LIMIT
```

```` javascript
db.collection.aggregate([
    { $limit: <positive integer> },
]);
````

Limits the number of documents passed to the next stage in the pipeline.

> ````javasc

Reviewer 2: ```suggestion
````javascript
```

db.collection.aggregate([
	{ $match: { price: 10 } },
    { $limit: 5 },
]);
````

Returns five results that matches the specified condition from the previous stage.

---

> ### $lookup

Reviewer 2: ```suggestion
### $LOOKUP
```

This operator allows the use of documents from another collection (table), similar to a **JOIN** from the SQL. The documents from the joined collection will be avaible through a new array.

There are two types of `$lookup`:

#### Equality Match

````javascript
db.collection.aggregate([
	{
        $lookup:
        {
          from: <collection to join>,
       	  localField: <field from the input documents>,
       	  foreignField: <field from the documents of the "from" collection>,
       	  as: <output array field>
        },
    },
]);
````

````javascript
db.collection.aggregate([
	{
        $lookup:
        {
          from: <inventory>,
       	  localField: <item>,
       	  foreignField: <sku>,
       	  as: <inventory_docs>
        },
    },
]);
````

----

> ### $group

Reviewer 2: ```suggestion
### $GROUP
```

This operator groups documents by the specified `_id` expression. It's worth mentioning that the specified `_id` is actually a "parameter" of `$group` and is not to be mistaken with `_id` from the collections.

> ````javas

Reviewer 2: ```suggestion
````javascript
```

db.collection.aggregate([
	{
		$group:
		{
			_id: <expression>,
			<field1>: { <accumulator1> : <expression1> },
			...
			<fieldN>: { <accumulatorN> : <expressionN> },
		},
	},
]);
````

The `<field>` may be named as you see fit, howerver, the `<accumulator>` must be a accumulator operator.

> ````javas

Reviewer 2: ```suggestion
````javascript
```

db.collection.aggregate([
  {
    $group : {
      _id : "$itemId",
      count: { $sum: 1},
    }
  }
]);
````

The output would be something like this:

````bash
{ _id : "Chair", count : 3 }
{ _id : "Table", count : 2 }
{ _id : "Lamp", count : 2 }
````

**OBS:** If you set `_id` as `null`, you will group all documents.

#### List of most used accumulators:

- $addToSet
- $avg
- $first
- $last
- $max
- $sum

---

> ### $unwind

Reviewer 2: ```suggestion
### $UNWIND
```

````javascript
db.collection.aggregate([
    { $unwind: <field path> },
]);
````

This operator desconstructs an array field into documents. To be more precisely, a document for each element of the array.

Let's take the following document as an example:

````javascript
{
    _id: 7,
    name: "GMusicApp",
    subscription: ["free", "premium", "family", "university"]
}
````

If we were to apply the `$unwind` operator:

````javascript
db.collection.aggregate([
    { $unwind: "$subscription" },
]);
````

We would get following output:

````javascript
{ _id: 7, name: "GMusicApp", subscription: "free" },
{ _id: 7, name: "GMusicApp", subscription: "premium" },
{ _id: 7, name: "GMusicApp", subscription: "family" },
{ _id: 7, name: "GMusicApp", subscription: "university" },
````

---

> ### $project

Reviewer 2: ```suggestion
### $PROJECT
```
